cmd/k8s-operator/logger.go

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !plan9
+
+package main
+
+import (
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	kzap "sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"tailscale.com/types/logger"
+)
+
+type (
+	logfSink struct {
+		logf logger.Logf
+	}
+)
+
+// wrapZapCore returns a zapcore.Core implementation that splits the core chain using zapcore.NewTee. This causes
+// logs to be simultaneously written to both the original core and the provided logger.Logf function.
+func wrapZapCore(core zapcore.Core, logf logger.Logf) zapcore.Core {
+	// We use a tee logger here so that logs are written to stdout/stderr normally while at the same time being
+	// sent upstream.
+	return zapcore.NewTee(core, zapcore.NewCore(&kzap.KubeAwareEncoder{
+		Encoder: zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()),
+		Verbose: true,
+	}, &logfSink{logf: logf}, zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
+		return true
+	})))
```

Comment:
This could do with some more line breaks to make it easier to read:

```suggestion
	return zapcore.NewTee(core,
		zapcore.NewCore(
			&kzap.KubeAwareEncoder{
				Encoder: zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()),
				Verbose: true,
			},
			&logfSink{
				logf: logf,
			},
			zap.DebugLevel,
		),
	)
```

Note I think you can also just use `zap.DebugLevel` instead of creating a struct for the level.

Are we sure we want `Verbose` for the `KubeAwareEncoder`? AFAIK I think we generally try to just print the names of objects and not full objects in logger fields. Docs:

```
// Verbose controls whether or not the full object is printed.
// If false, only name, namespace, api version, and kind are printed.
// Otherwise, the full object is logged.
```